### PR TITLE
Expose Thanos HTTP APIs to OpenAPI/protobuf Proposal

### PR DESCRIPTION
**Pull request by Hangzhi**
Signed-off-by: Hangzhi <1454678938@qq.com>

Related to : #4102 
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

### CODE REVIEW

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+| type     | title                                                        | status   | menu               |
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:1]: I think we use different format of front matter - hopefully we can get rid of this too, but right now it's required to be yaml between `---` letters.

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    - https://github.com/prometheus/prometheus/issues/7192
+    - https://github.com/prometheus/prometheus/issues/5567
+- Other docs:
+  - A proposal of LFX mentorship: https://docs.google.com/document/d/1GeyFCk4QkVSrx4pQ4OSmOA1KE_FvmJRi_aq9hr2GoZQ/edit?usp=sharing
+
+This design doc is proposing a design for Thanos HTTP API defined in protobuf and OpenAPI.
+
+## **Motivations**
+
+In order to improve Thanos usage for users, we would like to define our HTTP APIs in protobuf/OpenAPI and expose those in the website. This would allow users to use tools for documentation, validation, type checking, and even code generation to use our APIs efficiently.
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:23]:
```suggestion
In order to improve Thanos usage for users, we would like to define our HTTP APIs in protobuf/OpenAPI and expose those in our repository. This would allow users to use tools for documentation, validation, type checking, and even code generation to use our APIs efficiently.

Similarly we want reuse this work in Prometheus.

```

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+### **Pitfalls of the current solution**
+
+- Documentation needs to be written manually.
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:27]:
```suggestion
- Documentation, Server code and client code boilerplate needs to be written manually.
- It's hard to discover the current API programmatically.
- It's easy to make an accidental breaking changes when modifying API
```

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+## **Goals**
+
+- Define all APIs in protobuf.
+- Be able to generate OpenAPI3 from protobuf.
+- Auto-generate documentation with OpenAPI3 specification.
+- Generate server code from API specification (OpenAPI3 or protobuf).
+- Define all configuration potentially in protobuf too: https://github.com/openproto/protoconfig.
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:35]: Maybe we can drop that to focus more on APIs in this phase (:

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+- Optionally we would like to have them on the index page on every Thanos component server.
+
+## **Non-Goals**
+
+- Not define Thanos gRPC APIs in protobuf specification.
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:40]:
```suggestion
- Don't mix gRPC with HTTP APIs in same protobuf package

```

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+## **How**
+
+- Define Thanos APIs in protobuf specification.
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:44]:
```suggestion
- Define Thanos HTTP APIs in protobuf specification.

```

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+- Generate OpenAPI from protobuf with gnostic extension.
+- Generate documentation from OpenAPI with [swagger](https://github.com/swagger-api/swagger-codegen).
+- Generate server API client and server stubs potentially from OpenAPI with [swagger](https://github.com/swagger-api/swagger-codegen) or [oapi-codegen](https://github.com/deepmap/oapi-codegen).
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:47]:
```suggestion
- Generate server and client API stubs potentially from OpenAPI with [swagger](https://github.com/swagger-api/swagger-codegen) or [oapi-codegen](https://github.com/deepmap/oapi-codegen).

```

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+## **Alternatives**
+
+## **Only define HTTP RESTful API only in OpenAPI. Not in protobuf.**
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:51]:
```suggestion
## **Define HTTP RESTful API only in OpenAPI. Not in protobuf.**

```

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+1. Pros:
+   1. We are not sure whether we can definitely generate OpenAPI specifications from protobuf.
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:54]: * There are might be some complexity, edge case and extra tooling to make 3 step process (proto -> open API -> documentation to work

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+   1. We write API definitions in protobuf for consistency.
+   2. Protobuf definition is more concise.
+
+## **Define APIs in gRPC and have RESTful APIs alongside with [grpc - gateway](https://github.com/grpc-ecosystem/grpc-gateway)**
+
+1. Pros:
+   1. We can have gRPC and RESTful APIs at the same time.
+   2. We have gRPC APIs like rules API.
+2. Cons:
+   1. We don't know clearly how grpc-gateway works.
```
Comment (bwplotka) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:65]: * We need run another sidecar (complexity of running the system)
* Semantics of gRPC and HTTP might be different and surprising for end user.
* We want to reuse in Prometheus and Prometheus does not support gRPC (gRPC dependency was removed from codebase)
* Same port library is not maintained. (cmux)

**Review by bwplotka — DISMISSED** [review]
Nice, great work! I love it, LGTM 💪🏽

**Comment by ianbillett** [comment]
@Hangzhi `make check-docs` to see what is causing CI to fail.

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md (hunk 2 above)
Comment (ianbillett) [reply docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:23]: For readability it would be good to expand on the following parts:
* `In order to improve Thanos usage for users` - what do we mean by improve here? What positive benefit will this bring to users?  
* `protobuf/OpenAPI` - as far as I am aware these are two different & non-interchangeable things?

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md (hunk 3 above)
Comment (ianbillett) [reply docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:27]: I would reword the final point to be something like
> When modifying the API, it is very difficult to know if this breaks downstream users

Or similar :)

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md (hunk 4 above)
Comment (ianbillett) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:35]: IMO these are not goals _per-se_ but the steps we take to satisfy the goals - in fact, they are very similar to the `How` section below.

I think the goals here is to make the Thanos APIs easier to share, document and consume?

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md (hunk 7 above)
Comment (ianbillett) [reply docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:47]: I'm not a big fan of having 'potentially' in proposals - we're either going to do it or not :)

**Review by ianbillett — COMMENTED** [review]
Looks good - the content is good, but I think our motivations can be better explained.

**Comment by Hangzhi** [comment]
Hello @ianbillett. The test shows that the file is not formatted, but it's not clear how `not formatted`. And it results in errors in the e2e test, though this PR should not involve the e2e test 🤔 . Do you have any hints on how I can resolve these errors?

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+* It's hard to discover the current API programmatically.
+* When modifying the API, it is very difficult to know if this breaks downstream users. 
+
+## **Goals**
+
+* Define all APIs in protobuf.
+* Be able to generate OpenAPI3 from protobuf.
+* Auto-generate documentation with OpenAPI3 specification.
+* Generate server code from API specification (OpenAPI3 or protobuf).
+* Define all configuration potentially in protobuf too: https://github.com/openproto/protoconfig.
```
Comment (onprem) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:43]: I don't think this goal is in the scope of this proposal, let's remove this.

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+   2. We need to redefine our API service in gPRC since most of Thanos APIs and Prometheus APIs are RESTful APIs. 
+   3. We need to run another sidecar (complexity of running the system).
+   4. Semantics of gRPC and HTTP might be different and surprising for end user. 
+   5. We want to reuse in Prometheus and Prometheus does not support gRPC (gRPC dependency was removed from codebase).
+   6. Same port library is not maintained. (cmux)
+
+## **Action Plan**
+
+* [ ]  Experiments in rules API.
+* [ ]  Define all http APIs in Protobuf.
```
Comment (onprem) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:83]:
```suggestion
* [ ]  Define all HTTP APIs in Protobuf.

```

File: docs/proposals-accepted/202107-protobuf-openapi-httpapi.md [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+* [ ]  Generate documents, client code, and server stubs from OpenAPI specification.
```
Comment (onprem) [inline docs/proposals-accepted/202107-protobuf-openapi-httpapi.md:84]:
```suggestion
* [ ]  Generate documentation, client code, and server stubs from OpenAPI specification.

```

**Review by onprem — DISMISSED** [review]
Looking good to me, just some small non-blocker nits.

**Comment by ianbillett** [comment]
@Hangzhi 

Since you are not modifying the e2e tests we can safely conclude that failed due to a transient flake. You can trigger a re-test with an empty commit `git commit --allow-empty -m 'retest'`.


If you inspect the output of the [failing test](https://github.com/thanos-io/thanos/pull/4426/checks?check_run_id=3124716832) it shows you why the linter is failing. If you inspect the `makefile` that is running the tests, you will find the binary we run (`mdox`) which you can run locally to fix the issues. The command will be something like `mdox fmt -l --links.validate.config-file=./.mdox.validate.yaml docs/proposals/20201019-prometheus-rules-in-observatorium-api.md`

**Comment by Hangzhi** [comment]
@ianbillett  Thank you for your very helpful information! I finally got the error info in detail with ` mdox fmt -l docs/proposals-accepted/202107-protobuf-openapi-httpapi.md`.

**Review by bwplotka — APPROVED** [review]
Amazing work!

Let's merge it, it has all details we want, problem and goals outlined well 💪🏽 

On our 1:2 we acknowledged that this proposal has some risk - there are things we need to experiment with. This is fine though -we can iterate on it. Edit the "HOW" when we find new things or reject it at then end if needed. But current version shows the latest direction.
